addons/mail/tests/test_discuss_controller.py

```diff
@@ -207,3 +210,126 @@ def test_attachment_hijack(self):
+            self.env["mail.channel"].create_group(partners_to=[])["id"]
+        )
+        private_message = private_channel.message_post(body="private message")
+        fake_message = self.env["mail.message"].browse(private_message.id + 100)
+        public_w_guest = self.public_user
+        no_user = self.env["res.users"]  # same as public user but with no guest
+
+        def find_reactions(user, message, content):
+            if user == public_w_guest:
+                user_leaf = ("guest_id", "=", self.guest.id)
+            else:
+                user_leaf = ("partner_id", "=", (user or self.public_user).partner_id.id)
+            return self.env["mail.message.reaction"].search(
+                [("message_id", "=", message.id), ("content", "=", content), user_leaf]
+            )
+
+        def add_reaction(user, message, allowed):
+            res = self.opener.post(
+                url=f"{self.channel.get_base_url()}/mail/message/add_reaction",
+                json={
+                    "params": {
+                        "message_id": message.id,
+                        "content": "😊",
+                    },
+                },
+            )
+            reactions = find_reactions(user, message, "😊")
+            if allowed:
+                self.assertEqual(len(reactions), 1, "😊 should be added")
+            else:
+                self.assertEqual(len(reactions), 0, "😊 should not be added")
+            return json.loads(res.content)
+
+        def remove_reaction(user, message, allowed):
+            if message != fake_message:
+                message.with_user(user or self.public_user).sudo().with_context(
+                    guest=self.guest if user == public_w_guest else None
+                )._message_add_reaction("😀")
+            res = self.opener.post(
+                url=f"{self.channel.get_base_url()}/mail/message/remove_reaction",
+                json={
+                    "params": {
+                        "message_id": message.id,
+                        "content": "😀",
+                    },
+                },
+            )
+            reactions = find_reactions(user, message, "😀")
+            if allowed or message == fake_message:
+                self.assertEqual(len(reactions), 0, "😀 should be removed")
+            else:
+                self.assertEqual(len(reactions), 1, "😀 should not be removed")
+            return json.loads(res.content)
+
+        # (message, user, allowed)
+        subtests = (
+            (public_message, no_user, False),
+            (public_message, public_w_guest, False),
+            (public_message, self.user_portal, False),
+            (public_message, self.user_employee, True),
+            (public_message, self.user_demo, True),
+            (public_message, self.user_admin, True),
+            (invite_message, no_user, False),
+            (invite_message, public_w_guest, True),
+            (invite_message, self.user_portal, True),
```

Comment:
Note: this is effectively the only line that fails before this PR (both add/remove reaction).